File: src/learn/arrays-hashing/easy/validAnagram.ts
const base = "a".charCodeAt(0);

function isAnagram(s: string, t: string): boolean {
    // P 
    //  - two strings =>  (lowercase english letters a ~ z)
    //  - for each string, 1 <= length <= 5 * 10^4
    // R
    //  - Boolean => whether if it is anagram
    // E
    //  - isAnagram("anagram", "nagaram") = true
    //  - isAnagram("rat", "car") = false
    //  - isAnagram("car", "rac") = true

    if (s.length !== t.length) return false;

    const p = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101];

    let totalS = 1;
    let totalT = 1;

    for (let i = 0; i < s.length; i++) {
        totalS *= p[s.charCodeAt(0) - 97];
        totalT *= p[t.charCodeAt(0) - 97];
    }

    return totalS === totalT;
}

console.log(isAnagram("anagram", "nagaram"));
console.log(isAnagram("rat", "car"));

File: src/solution/arrays-hashing/valid_anagram.go
package main

func IsAnagram(s string, t string) bool {
    /*
    1 <= s.length, t.length <= 5 * 104
    s and t consist of lowercase English letters (Important)
    */
    if len(s) != len(t) {
        return false
    }

    a := make([]int, 26, 26)
    b := make([]int, 26, 26)

    for i := 0; i < len(s); i++ {
        a[s[i] - 'a'] += 1
        b[t[i] - 'a'] += 1
    }

    for i := 0; i < len(a); i++ {
        if a[i] != b[i] {
            return false
        }
    }

    return true
}

var prime_encoding = [26]int{2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,
61,67,71,73,79,83,89,97,101}

func IsAnagramAlt1(s string, t string) bool {
    if len(s) != len(t) {
        return false
    }

    a := 0; b := 0

    for i := 0; i < len(s); i++ {
        a *= prime_encoding[s[i] - 'a']
        b *= prime_encoding[t[i] - 'a']
    }

    return a == b
}
